index.d.ts

```diff
@@ -0,0 +1,77 @@
+// Type definitions for express-session 1.16
+// Project: https://github.com/expressjs/session
+// Definitions by: Hiroki Horiuchi <https://github.com/horiuchi>
+//                 Jacob Bogers <https://github.com/jacobbogers>
+//                 Naoto Yokoyama <https://github.com/builtinnya>
+//                 Ryan Cannon <https://github.com/ry7n>
+//                 HoldYourWaffle <https://github.com/HoldYourWaffle>
+
+import express from 'express'
+import node from 'events'
+
+
+
+declare global {
+  namespace Express {
+    interface Request { //inject additional properties on express.Request
+      session?: Session;
+      sessionID?: string;
+    }
+  }
+
+  interface SessionData {
+    //seperate interface to allow application-specific declaration merging (see README)
+  }
+}
+
+
+
+interface SessionOptions {
+  secret: string | string[];
+
+  genid?(req: express.Request): string;
+  name?: string;
+  store?: Store;
+  cookie?: express.CookieOptions;
+  rolling?: boolean;
+  resave?: boolean;
+  proxy?: boolean;
+  saveUninitialized?: boolean;
+  unset?: string;
+}
+
+declare function session(options?: SessionOptions): express.RequestHandler;
+export default session;
+
+
+
+interface Session extends SessionData {
+  id: string;
+  cookie: SessionCookie;
+
+  regenerate(callback: (err: any) => void): Session;
+  destroy(callback: (err: any) => void): Session;
+  reload(callback: (err: any) => void): Session;
+  save(callback?: (err: any) => void): Session;
+  touch(): Session;
+}
+
+declare interface SessionCookie extends express.CookieOptions { //XXX actually a class
```

Comment:
what does the XXX comment here mean? it is a todo?

Reply:
It's a note, mostly targeted at the original authors. The issue is that `SessionCookie` (which is a reference to the 'class' defined in [cookie.js](https://github.com/expressjs/session/blob/master/session/cookie.js)) is technically a class, but as far as I can tell it shouldn't be referenced directly by the user. This is mainly what I meant by not understanding the cookie system, I was hoping the original authors could shine some light on what they decided to do and why.

(@horiuchi @jacobbogers @builtinnya @Ry7n)

Reply:
I mean, it is exported as `require('express-session').Cookie`. It's needed by stores created without the use of the `require('express-session').Store` export.

Reply:
I have now defined the hierarchy as follows:
- There's a class `Cookie`
- This class implements `CookieOptions`, meaning it has all the properties defined by it
- The interface (basically a list of properties) `CookieOptions` is based on the README and your comments above
- `Session.cookie` has the type `Cookie`

Does this sound correct to you?

Reply:
Yea